flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java

```diff
@@ -340,6 +362,65 @@ public void resetNFAChanged() {
 		return Tuple2.of(result, timeoutResult);
 	}
 
+	private void discardComputationStatesAccordingToStrategy(Queue<ComputationState<T>> computationStates,
```

Comment:
This method is very fragile as it makes the assumption that the different patterns appear in the match in the order that they were declared in the pattern declaration, e.g. for a pattern `A -> B -> C` it assumes that the partial matches will appear in the order [`matches for A`, `matches for B`, `matches for C`]. This is *NOT* correct as we use a `HashMap` which does not provide any ordering guarantees. 

A quick fix would be to use a `LinkedHashMap` as the return type for the `extractCurrentMatches` and put the results there in the correct order.